### DISCUSSION TITLE
Sputnik Upgrade options

### DISCUSSION THREAD

**Discussion by TrevorJTClarke** [near-daos/sputnik-dao-contract#170] [Polls]
As of neard 1.26.0, there is a protocol change to gas calculations.
Sputnik V2 utilized hardcoded gas amounts, which are now insufficient to cover execution of upgrades.

1.26.0 is scheduled to ship May 18th.

Here's the options available:

### 1. New V3 Factory & DAOs

* Create a new factory, owned by a DAO, shipping V3 code as a starting point.
* Any DAO can deploy a new DAO instance in V3, and migrate their funds
* No V3 DAOs will have FA keys
* All upgrades to V3 Factory & DAOs are done via Factory DAO following new Upgrade flows

### 2. V2 -> V3 Conversion with some locked DAOs

* Attempt to help large & active DAOs to upgrade pre-protocol change. This would require a massive marketing/outreach program.
* Could potentially make future use of V2 DAOs harder

#### 2.A. Speed upgrade pre-1.26.0 release

* Needs to have audit complete, factory owner onboarded into needs, DAOs upgrading pre-May 18th

#### 2.B. Forceful upgrade of all FA key DAOs

* Same as A. but doesn't respect the sanctity of the DAOs that still have FullAccess Keys
#### 2.C. Upgrade ONLY Factory to pause future DAO creation

* Good outcome if Option 1. is happening.

### 3. Delay or Remove Protocol Gas changes

* Lobby for validators & protocol core team to roll back gas changes until further DAO upgrades (and any other contracts affected) have time to ship code changes.
* Removal would allow 1.26.0 release to ship on-time
* Delay would ultimately not solve any core issues with the contract, but allow for a greater amount of DAOs to upgrade.

**Comment by starpause** [comment]
New V3 factory seems the cleanest and most feasible.

Could the forced upgrade (option 2) be done with ONLY the updated gas amounts? If that was feasible it would mitigate the amount of outreach required. We would only need to reach the DAOs that have removed their FA keys to ensure they got that upgrade installed before May 18th.

**Reply by TrevorJTClarke** [reply]
Forced upgrade on Option 2 is still possible, however I believe it still violates trust of the DAO. Imagine i deploy a malicious DAO code without your knowledge, not good.

It also would need to be done pre-protocol change, very difficult timing.

Since it excludes the non-FA key DAOs, i remain sold on doing a diff option to cover all DAOs with the same solution

**Comment by bowenwang1996** [comment]
Adding more context here after I talked to Trevor:
* Sputnik v2 factory has hardcoded gas amount for contract upgrades that would not be sufficient after the 1.26.0 release
* DAOs created through sputnik v2 have that hardcoded gas limit. As a result, they won't be able to upgrade through the upgrade function after the 1.26.0 release
* Some DAOs have full access keys to the contract. They won't be affected by the release as they could simply bypass the upgrade function and deploy a new contract directly.
* Some DAOs with a large amount of collateral such as Aurora and Ref finance, however, deleted full access keys to their dao contracts. As a result, they won't be able to upgrade the contract after the 1.26.0 release altogether.
* Those DAOs can still manually migrate to v3 by first creating a new dao contract and then transferring collaterals to the new contract.

If the large DAOs are okay with manual migration to v3 I think we should proceed with option 1.
